src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotificationStrategy.java

```diff
@@ -0,0 +1,97 @@
+                return "continuous-integration/jenkins/pr-merge";
+            } else {
+                return "continuous-integration/jenkins/pr-head";
+            }
+        } else {
+            return "continuous-integration/jenkins/branch";
+        }
+    }
+
+    protected String generateUrl() {
+        String url = null;
+        try {
+            if (null != build) {
+                url = DisplayURLProvider.get().getRunURL(build);
+            }
+            else {
+                url = DisplayURLProvider.get().getJobURL(job);
+            }
+        } catch (IllegalStateException e) {
+            listener.getLogger().println(
+                    "Can not determine Jenkins root URL. Commit status notifications are disabled "
+                            + "until a root URL is"
+                            + " configured in Jenkins global configuration.");
+        }
+        return url;
+    }
+
+    protected String generateMessage() {
+        if (null != build) {
+            Result result = build.getResult();
+            if (Result.SUCCESS.equals(result)) {
+                return Messages.GitHubBuildStatusNotification_CommitStatus_Good();
+            } else if (Result.UNSTABLE.equals(result)) {
+                return Messages.GitHubBuildStatusNotification_CommitStatus_Unstable();
+            } else if (Result.FAILURE.equals(result)) {
+                return Messages.GitHubBuildStatusNotification_CommitStatus_Failure();
+            } else if (Result.ABORTED.equals(result)) {
+                return Messages.GitHubBuildStatusNotification_CommitStatus_Aborted();
+            } else if (result != null) { // NOT_BUILT etc.
+                return Messages.GitHubBuildStatusNotification_CommitStatus_Other();
+            } else {
+                return Messages.GitHubBuildStatusNotification_CommitStatus_Pending();
+            }
+        }
+        return Messages.GitHubBuildStatusNotification_CommitStatus_Queued();
+    }
+
+    protected GHCommitState generateState() {
+        if (null != build) {
+            Result result = build.getResult();
+            if (Result.SUCCESS.equals(result)) {
+                return GHCommitState.SUCCESS;
+            } else if (Result.UNSTABLE.equals(result)) {
+                return GHCommitState.FAILURE;
+            } else if (Result.FAILURE.equals(result)) {
+                return GHCommitState.ERROR;
+            } else if (Result.ABORTED.equals(result)) {
+                return GHCommitState.ERROR;
+            } else if (result != null) { // NOT_BUILT etc.
+                return GHCommitState.ERROR;
+            }
+        }
+        return GHCommitState.PENDING;
+    }
+
+    public final GitHubBuildStatusNotificationBundle createNotification(Job<?, ?> job, Run<?, ?> build, TaskListener listener,
+                                                                        SCMSource source, SCMHead head) {
+        this.job = job;
```

Comment:
I wouldn't store these as fields, it will cause race conditions, etc.

Put all the details into a `GitHubNotificationRequest` immutable class and then just pass the request and the `TaskListener` as parameters to `generateContext(GitHubNotificationRequest, TaskListener)`, `generateUrl(GitHubNotificationRequest, TaskListener)` etc...

Reply:
Doesn't have to be `GitHubNotificationRequest` could be `GitHubNotificationContext` if you feel that makes better sense... and then `GitHubNotificationDetails` could be `GitHubNotificationRequest`

So the `GitHubNotificationStrategy` can decide, for each context, what notification requests should be made, then the plugin goes and makes those.

Hmm... I wonder would it make sense to allow for multiple strategies? That way a trait can add in the strategies and we don't need to worry about multiple traits trying to set a strategy they want and we use the disable notification to be a disable default notification strategy... and (for the win) it retains semantic compatibility...

Reply:
I thought about multiple strategies, but wasn't sure how to resolve when two different strategies implement the same component.

Reply:
We could keep track and only let the first one actually notify... or report an error on the second

Reply:
We could keep track and only let the first one actually notify... or report an error on the second

Reply:
Working on multiple strategies, my feeling was that multiple strategies could together produce a single notification (one contributing a URL and another a context for example). This isn't gelling well with the idea of a strategy being able to itself produce multiple notifications. Will take a bit of thinking to work out how various counts of notification from different strategies will combine together.

Reply:
Happy to see where your experiment takes you, but I think it will be a mess.. much better to let each strategy contribute the notifications it wants and just apply them all (if the user configures overlapping... let the last one win!!! (user error) in the KISS principle)

OTOH If you have a grand vision that allows for co-operation between traits, let's see where that goes

Reply:
Yep, simplicity is probably for the best here. If I have an epiphany I might revisit it later.